### PR TITLE
Platform and Broker API

### PR DESCRIPTION
**Pull request by NickyMateev**
Represents an implementation of the Broker API and Platform API.

- [x] Platform REST API implementation
- [x] Broker REST API implementation
- [x] Code refactoring and cleanup
- [x] Tests

### CODE REVIEW

File: test/api/api_test.go [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+	"net/http/httptest"
+	"os"
+	"testing"
+
+	"github.com/Peripli/service-manager/bootstrap"
+	"github.com/Peripli/service-manager/env"
+	"github.com/gorilla/mux"
+	"github.com/sirupsen/logrus"
+
+	"github.com/gavv/httpexpect"
```
Comment (dpanayotov) [inline test/api/api_test.go]: this is not listed as a dependency

File: test/api/api_test.go [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+	}
+	return srv.Router
+}
+
+var _ = Describe("Service Manager API", func() {
+	var testServer *httptest.Server
+
+	BeforeSuite(func() {
+		testServer = httptest.NewServer(getServerRouter())
+		SM = httpexpect.New(GinkgoT(), testServer.URL)
```
Comment (dpanayotov) [inline test/api/api_test.go]: this is hard to follow, maybe name this file api_suite_test, move the util things here and let testBrokers and testPlatforms be in the broker/platforms_test

File: .travis.yml [hunk 3]
```diff
@@ -16,14 +16,17 @@ install:
...
+services:
+  - postgresql
+
 script:
   # Execute static checks
   - gometalinter --vendor ./...
 
   # Execute tests
-  - go test $(go list ./... | grep -v "..fakes") -covermode=count -coverprofile=profile.cov
+  - go test ./... -coverpkg $(go list ./... | egrep -v "fakes|test" | paste -sd "," -) -coverprofile=profile.cov
```
Comment (dpanayotov) [inline .travis.yml]: this is starting to look like it needs explanation

File: api/broker/broker.go [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+// Routes returns slice of routes which handle broker operations
+func (brokerCtrl *Controller) Routes() []rest.Route {
+	return []rest.Route{
+		{
+			Endpoint: rest.Endpoint{
+				Method: http.MethodPost,
+				Path:   url,
+			},
+			Handler: rest.APIHandler(brokerCtrl.addBroker),
```
Comment (dpanayotov) [inline api/broker/broker.go]: POST to /v1/service_brokers/ returns all service brokers. If I remove the slash '/' then it's ok.

File: api/broker/broker-controller.go [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+	brokerID := mux.Vars(request)["broker_id"]
+	brokerStorage := brokerCtrl.BrokerStorage
+	broker, err := brokerStorage.Get(brokerID)
+	if err != nil {
+		if err == storage.ErrNotFound {
+			return rest.CreateErrorResponse(err, http.StatusNotFound, "NotFound")
+		}
+		return err
+	}
+	return rest.SendJSON(response, http.StatusOK, broker)
```
Comment (dpanayotov) [inline api/broker/broker-controller.go]: the Location header should be included

File: storage/postgres/storage.go [hunk 6]
```diff
@@ -47,17 +50,72 @@ func (storage *postgresStorage) Broker() storage.Broker {
...
 func (storage *postgresStorage) Close() error {
 	return storage.db.Close()
 }
+
+func updateSchema(db *sqlx.DB) error {
+	driver, err := migratepg.WithInstance(db.DB, &migratepg.Config{})
+	if err != nil {
+		return err
+	}
+	m, err := migrate.NewWithDatabaseInstance("file://storage/postgres/migrations", "postgres", driver)
```
Comment (dpanayotov) [inline storage/postgres/storage.go]: this file loading doesn't work if you import the service manager in a platform-specific project

File: storage/postgres/platform.go [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+	set = append(set, "updated_at = :updated_at")
+	return fmt.Sprintf("UPDATE %s SET %s WHERE id = :id", platformTable, strings.Join(set, ", "))
+}
+
+func checkUniqueViolation(err error) error {
+	if err == nil {
+		return nil
+	}
+	sqlErr, ok := err.(*pq.Error)
+	if ok && sqlErr.Code.Name() == "unique_violation" {
```
[the remark this answers is not in the document]
Comment (pankrator) [reply storage/postgres/platform.go]: This value comes from the Postgres database, not the driver itself, so I think it is safe to use.

File: storage/postgres/broker.go [hunk 8]
```diff
@@ -17,37 +17,170 @@
...
+		return nil
+	})
+}
+
+func (store *brokerStorage) Update(broker *rest.Broker) error {
+	return transaction(store.db, func(tx *sqlx.Tx) error {
+		updateQueryString := generateUpdateQueryString(broker)
+
+		brokerDTO := convertBrokerToDTO(broker)
+		if updateQueryString != "" {
```
[the remark this answers is not in the document]
Comment (dotchev) [reply storage/postgres/broker.go]: it can, see https://github.com/Peripli/service-manager/pull/35/files#diff-3e19e4e3bf8217b12e8453d9a73db40fR167

File: storage/postgres/model.go [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+	platformTable = "platforms"
+
+	// table db table name for credentials
+	credentialsTable = "credentials"
+
+	basicCredentialsType = 1
+)
+
+// Credentials dto
+type Credentials struct {
```
[the remark this answers is not in the document]
Comment (dotchev) [reply storage/postgres/model.go]: I don't get your point. These are the Postgres-specific structures.

File: storage/postgres/platform.go [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+	"github.com/lib/pq"
+)
+
+type platformStorage struct {
+	db *sqlx.DB
+}
+
+var (
+	// selectByID selects platform by id
+	selectByID = "SELECT id, type, name, description, created_at, updated_at FROM " + platformTable + " WHERE id=$1"
```
[the remark this answers is not in the document]
Comment (dotchev) [reply storage/postgres/platform.go]: Yes, it is clear they were written by different people.

I would prefer having the SQL in the code that uses it, then it is easier to read it
In this particular case you don't see what is mapped to `$1`.

File: storage/postgres/platform.go [hunk 11]
```diff
@@ -0,0 +1,165 @@
...
+		if err != nil {
+			return err
+		}
+		return checkRowsAffected(result)
+	})
+}
+
+func (storage *platformStorage) Update(platform *rest.Platform) error {
+	updateQuery := platformUpdateQueryString(platform)
+	if updateQuery == "" {
```
[the remark this answers is not in the document]
Comment (dotchev) [reply storage/postgres/platform.go]: see https://github.com/Peripli/service-manager/pull/35/files#diff-574668b8a221bcfea851affe6efd1475R149

File: storage/postgres/platform.go [hunk 12]
```diff
@@ -0,0 +1,165 @@
...
+		return err
+	}
+	if rowsAffected != 1 {
+		return store.ErrNotFound
+	}
+	return nil
+}
+
+func platformUpdateQueryString(platform *rest.Platform) string {
+	set := make([]string, 0, 4)
```
[the remark this answers is not in the document]
Comment (dotchev) [reply storage/postgres/platform.go]: `make` returns a slice, see https://golang.org/pkg/builtin/#make
here we specify the capacity to avoid reallocation

File: storage/postgres/platform.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (dotchev) [reply storage/postgres/platform.go]: there are also error codes, but we chose the name as it is more descriptive
also these error names are defined by Postgres, so they are unlikely to change
https://www.postgresql.org/docs/devel/static/errcodes-appendix.html

File: storage/postgres/storage.go (hunk 6 above)
Comment (dotchev) [reply storage/postgres/storage.go]: What do you mean by "platform-specific project"? I am not sure we are on the same page here.
I expect SM to be deployed on different platforms (CF, K8S) only by changing the configuration via env vars.

Specifically for migrations I don't expect to use anything else than files. SM should package these migrations.

File: storage/postgres/storage.go [hunk 13]
```diff
@@ -47,17 +50,72 @@ func (storage *postgresStorage) Broker() storage.Broker {
...
+}
+
+func transaction(db *sqlx.DB, f func(tx *sqlx.Tx) error) error {
+	tx, err := db.Beginx()
+	if err != nil {
+		logrus.Error("Could not create transaction")
+		return err
+	}
+	ok := false
+	defer func() {
```
[the remark this answers is not in the document]
Comment (dotchev) [reply storage/postgres/storage.go]: `defer` is needed in case `f` panics

File: test/api/api_test.go [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+	})
+
+	AfterSuite(func() {
+		if testServer != nil {
+			testServer.Close()
+		}
+	})
+
+	Describe("Service Brokers", testBrokers)
```
[the remark this answers is not in the document]
Comment (dotchev) [reply test/api/api_test.go]: to reduce file size and make it more manageable

File: test/api/api_test.go (hunk 2 above)
Comment (dotchev) [reply test/api/api_test.go]: util things were here initially but this caused cyclic deps

File: api/broker/broker.go (hunk 4 above)
Comment (pankrator) [reply api/broker/broker.go]: See: http://www.gorillatoolkit.org/pkg/mux#Router.StrictSlash

File: api/platform/platform-controller.go [hunk 15]
```diff
@@ -0,0 +1,175 @@
...
+	err := platformStorage.Update(newPlatform)
+	if err != nil {
+		if err == storage.ErrUniqueViolation {
+			return rest.CreateErrorResponse(err, http.StatusConflict, "Conflict")
+		} else if err == storage.ErrNotFound {
+			return rest.CreateErrorResponse(errorMissingPlatform(platformID), http.StatusNotFound, "NotFound")
+		}
+		return err
+	}
+	platform, err := platformStorage.Get(platformID)
```
[the remark this answers is not in the document]
Comment (KirilKabakchiev) [reply api/platform/platform-controller.go]: the idea is to do as least db calls as possible - if we can avoid the call  I think we should go for it  right?

File: rest/types.go [hunk 16]
```diff
@@ -0,0 +1,97 @@
...
+	Description string       `json:"description"`
+	CreatedAt   time.Time    `json:"created_at"`
+	UpdatedAt   time.Time    `json:"updated_at"`
+	BrokerURL   string       `json:"broker_url"`
+	Credentials *Credentials `json:"credentials,omitempty"`
+}
+
+// MarshalJSON override json serialization for http response
+func (b *Broker) MarshalJSON() ([]byte, error) {
+	type B Broker
```
[the remark this answers is not in the document]
Comment (KirilKabakchiev) [reply rest/types.go]: I think it would be better to just drop the method overrall and add nanoseconds in the specification - simpler for everybody , right?

File: rest/types.go [hunk 17]
```diff
@@ -0,0 +1,97 @@
...
+		*P
+	}{
+		P:         (*P)(p),
+		CreatedAt: util.ToRFCFormat(p.CreatedAt),
+		UpdatedAt: util.ToRFCFormat(p.UpdatedAt),
+	})
+}
+
+// Broker broker struct
+type Broker struct {
```
[the remark this answers is not in the document]
Comment (KirilKabakchiev) [reply rest/types.go]: registering a broker would have a parameter for spaceguid

File: api/broker/broker.go (hunk 4 above)
Comment (dpanayotov) [reply api/broker/broker.go]: I know that. My thought was that this should be marked as an issue and be fixed soonish

File: storage/postgres/storage.go (hunk 6 above)
Comment (dpanayotov) [reply storage/postgres/storage.go]: By platform-specific project I mean "service-manager-cf" for example. It will have service-manager as dependency and the only thing that it will do is override some environment variables.
Then the service-manager-cf will be deployed on CF. So having the service-manager as dependency introduces this issue where files are not looked for in the dependency but in the wrapper

File: api/broker/broker-controller.go [hunk 18]
```diff
@@ -0,0 +1,132 @@
...
+
+func (brokerCtrl *Controller) addBroker(response http.ResponseWriter, request *http.Request) error {
+	logrus.Debugf("POST %s", request.RequestURI)
+
+	broker := rest.Broker{}
+	if err := rest.ReadJSONBody(request, &broker); err != nil {
+		return err
+	}
+
+	if broker.Name == "" || broker.BrokerURL == "" || broker.Credentials == nil {
```
[the remark this answers is not in the document]
Comment (dpanayotov) [reply api/broker/broker-controller.go]: Not only empty, if the are not present the request is accepted:
```json
{
	"name": "test",
	"broker_url": "http://example.com",
	"credentials": {
		"basic": {
		}	
	}
}
```

File: storage/postgres/broker.go [hunk 19]
```diff
@@ -17,37 +17,170 @@
...
 
 type brokerStorage struct {
 	db *sqlx.DB
 }
 
-func (storage *brokerStorage) Create(ctx context.Context, broker *types.Broker) error {
-	return nil
+type fetcher func(dest interface{}, query string, args ...interface{}) error
+
+func (store *brokerStorage) Create(broker *rest.Broker) error {
```
[the remark this answers is not in the document]
Comment (dotchev) [reply storage/postgres/broker.go]: the storage api provides a common interface to the persistency layer so it has to use a common model, currently this is the rest model
Note, we don't aim for the storage api ro be reusable in other contexts

File: storage/postgres/broker.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (dpanayotov) [reply storage/postgres/broker.go]: My bad :)

File: rest/types.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (dotchev) [reply rest/types.go]: we can make this a configuration of the CF proxy to register brokers locally instead of globally

File: storage/postgres/broker.go (hunk 19 above)
[the remark this answers is not in the document]
Comment (dpanayotov) [reply storage/postgres/broker.go]: Well, the storage needs to return everything from the database and doesn't need to know about how everyone else will use this data. For example, this breaks the "call service broker" functionality written here https://github.com/Peripli/service-manager/pull/46/files#diff-404f10463ebdd7100ab15d74f6763e45R197 as the credentials are not returned by the storage because it is concerned with another layer.

[the remark this answers is not in the document]
Comment (KirilKabakchiev) [reply storage/postgres/broker.go]: yeah, I think we discussed this when we talked about eager / lazy fetching of related tables - this here should be doing lazy fetching probably

File: storage/postgres/broker.go [hunk 20]
```diff
@@ -17,37 +17,170 @@
...
+			if err != nil {
+				return err
+			}
+			if affectedRows != 1 {
+				return storage.ErrNotFound
+			}
+		}
+
+		if broker.Credentials != nil {
+			credentialsDTO := convertCredentialsToDTO(broker.Credentials)
```
[the remark this answers is not in the document]
Comment (georgifarashev) [reply storage/postgres/broker.go]: Will be checked in the controller

File: rest/types.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (georgifarashev) [reply rest/types.go]: Currently this is not added in the broker

File: api/platform/platform-controller.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (georgifarashev) [reply api/platform/platform-controller.go]: If we return directly it will not be the whole platform

File: api/common/common.go [hunk 21]
```diff
@@ -0,0 +1,61 @@
...
+func CheckErrors(errors ...error) error {
+	for _, err := range errors {
+		if err != nil {
+			_, ok := err.(*types.ErrorResponse)
+			if ok {
+				return err
+			}
+		}
+	}
+	return errors[0]
```
Comment (dpanayotov) [inline api/common/common.go]: would panic if errors is not proviced

File: storage/postgres/broker.go [hunk 22]
```diff
@@ -17,37 +17,156 @@
...
-		Password: "brokerAdmin",
-	}, nil
+func (bs *brokerStorage) GetAll() ([]types.Broker, error) {
+	brokers := []Broker{}
+	err := bs.db.Select(&brokers, "SELECT id, name, description, created_at, updated_at, broker_url FROM "+brokerTable)
+	if err != nil {
+		logrus.Error("An error occurred while retrieving all brokers")
+		return nil, err
+	}
+	restBrokers := make([]types.Broker, len(brokers))
```
Comment (dpanayotov) [inline storage/postgres/broker.go]: note for future improvement - if this remains as is, the len should actually be cap.

File: test/api/api_test.go (hunk 1 above)
Comment (dpanayotov) [reply test/api/api_test.go]: Still not listed in Gopkg.tolm see it

Comment (pankrator) [reply test/api/api_test.go]: It doesn't need to be in Gopkg.toml. It is in the lock file. Most of the dependencies are only in the lock file

File: rest/types.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (KirilKabakchiev) [reply rest/types.go]: what about this one? can we remove?

File: test/api/api_test.go (hunk 1 above)
Comment (dpanayotov) [reply test/api/api_test.go]: This way, the last commit from the master branch will be taken. I'd rather not have `branch = "master"`, but exact revision number for a guaranteed reproducible build.
Sure, this will work, but it's not good practice

**Comment by KirilKabakchiev** [comment]
Squash before merging
